File: leetcode_ly/c++/Array/GameOfLife.cpp
class Solution {
public:
    void gameOfLife(vector<vector<int>>& board) {
        int m = board.size();
        int n = board[0].size();
        
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                int num = getLiveNum(board,i,j);
                if(board[i][j] == 1){
                    if(num < 2 || num > 3) board[i][j] = 2;
                }
                else{
                    if(num == 3) board[i][j] = 3;
                }
            }
        }
        
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                board[i][j] %= 2;
            }
        }
    }
    
    int getLiveNum(vector<vector<int>>& board,int i,int j){
        int row = i - 1;
        int col = j - 1;
        int count = 0;
        for(int p = row; p < row + 3; p++){
            for(int q = col; q < col + 3; q++){
                if(p >= 0 && p < board.size() && q >= 0 && q < board[0].size()){
                    if(board[p][q] == 1 || board[p][q] == 2) count++;
                }
            }
        }
        count = count - board[i][j];
        return count;//把自己也算进去了
    }
};

File: leetcode_ly/java/Array/GameOfLife.java
public class Solution {
    public void gameOfLife(int[][] board) {
        int m = board.length;
        if(m == 0) return;
        int n = board[0].length;
        
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(board[i][j] == 1){
                    if(getNum(i,j,board) < 2) board[i][j] = 3;
                    else if(getNum(i,j,board) > 3) board[i][j] = 2;
                    else board[i][j] = 4;
                }
                else if(board[i][j] == 0){
                    if(getNum(i,j,board) == 3) board[i][j] = 5;
                }
            }
        }
        
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(board[i][j] == 4 || board[i][j] == 5) board[i][j] = 1;
                else board[i][j] = 0;
            }
        }
        return;
    }
    
    public int getNum(int i,int j,int[][] board){
        int m = board.length;
        int n = board[0].length;
        int count = 0;
        for(int p = i - 1; p < i + 2; p++){
            for(int q = j - 1; q < j + 2; q++){
                if(p >= 0 && p < m && q >= 0 && q < n){
                    if(p == i && q == j) continue;
                    else{
                        if(board[p][q] == 1 || board[p][q] == 2 || board[p][q] == 3 || board[p][q] == 4){
                            count++;
                        }
                    }
                }
            }
        }
        return count;
    }
}